desktop/notification/fdo_test.go

```diff
@@ -0,0 +1,611 @@
+				Headers: map[dbus.HeaderField]dbus.Variant{
+					dbus.FieldReplySerial: dbus.MakeVariant(msg.Serial()),
+					dbus.FieldSender:      dbus.MakeVariant(":1"), // This does not matter.
+					// dbus.FieldDestination is provided automatically by DBus test helper.
+					dbus.FieldSignature: dbus.MakeVariant(responseSig),
+				},
+				Body: []interface{}{uint32(7)},
+			}
+			return []*dbus.Message{response}, nil
+		}
+		return nil, fmt.Errorf("unexpected message #%d: %s", n, msg)
+	})
+	id, err := srv.SendNotification(&notification.Message{
+		AppName:       "app-name",
+		Icon:          "icon",
+		Summary:       "summary",
+		Body:          "body",
+		ExpireTimeout: 1000,
+		ReplacesID:    notification.ID(42),
+		Actions: []notification.Action{
+			{ActionKey: "key-1", LocalizedText: "text-1"},
+			{ActionKey: "key-2", LocalizedText: "text-2"},
+		},
+		Hints: []notification.Hint{
+			{Name: "hint-str", Value: "str"},
+			{Name: "hint-bool", Value: true},
+		},
+	})
+	c.Assert(err, IsNil)
+	c.Check(id, Equals, notification.ID(7))
+}
+
+func (s *fdoSuite) TestSendNotificationError(c *C) {
+	srv := s.connectWithHandler(c, func(msg *dbus.Message, n int) ([]*dbus.Message, error) {
+		switch n {
+		case 0:
+			s.checkNotifyRequest(c, msg)
+			response := s.nameHasNoOwnerResponse(c, msg)
+			return []*dbus.Message{response}, nil
+		}
+		return nil, fmt.Errorf("unexpected message #%d: %s", n, msg)
+	})
+	_, err := srv.SendNotification(&notification.Message{})
+	c.Assert(err, ErrorMatches, "org.freedesktop.DBus.Error.NameHasNoOwner")
+}
+
+func (s *fdoSuite) TestCloseNotificationSuccess(c *C) {
+	srv := s.connectWithHandler(c, func(msg *dbus.Message, n int) ([]*dbus.Message, error) {
+		switch n {
+		case 0:
+			s.checkCloseNotificationRequest(c, msg)
+			c.Check(msg.Body[0], Equals, uint32(42))
+			response := &dbus.Message{
+				Type: dbus.TypeMethodReply,
+				Headers: map[dbus.HeaderField]dbus.Variant{
+					dbus.FieldReplySerial: dbus.MakeVariant(msg.Serial()),
+					dbus.FieldSender:      dbus.MakeVariant(":1"), // This does not matter.
+					// dbus.FieldDestination is provided automatically by DBus test helper.
+				},
+			}
+			return []*dbus.Message{response}, nil
+		}
+		return nil, fmt.Errorf("unexpected message #%d: %s", n, msg)
+	})
+	err := srv.CloseNotification(notification.ID(42))
+	c.Assert(err, IsNil)
+}
+
+func (s *fdoSuite) TestCloseNotificationError(c *C) {
+	srv := s.connectWithHandler(c, func(msg *dbus.Message, n int) ([]*dbus.Message, error) {
+		switch n {
+		case 0:
+			s.checkCloseNotificationRequest(c, msg)
+			response := s.nameHasNoOwnerResponse(c, msg)
+			return []*dbus.Message{response}, nil
+		}
+		return nil, fmt.Errorf("unexpected message #%d: %s", n, msg)
+	})
+	err := srv.CloseNotification(notification.ID(42))
+	c.Assert(err, ErrorMatches, "org.freedesktop.DBus.Error.NameHasNoOwner")
+}
+
+type testObserver struct {
+	notificationClosed func(notification.ID, notification.CloseReason) error
```

Comment:
(nitpick) I think I would have done something like:
```
type testObserver struct {
	notificationClosed []struct {
		notification.ID
		notification.CloseReason
	}
	notificationClosedErr error

	actionInvoked []struct {
		notification.ID
		string
	}
	actionInvokedErr error
}

func (o *testObserver) NotificationClosed(id notification.ID, reason notification.CloseReason) error {
	o.notificationClosed = append(o.notificationClosed, struct {
		notification.ID
		notification.CloseReason
	}{id, reason})

	return o.notificationClosedErr
}
...
```
and then just "c.Check(DeepEquals)" on the testObserver.notificationClosed. Slightly more compact in the individual tests but probably just a matter of preferences.